File: DMOJ/GFSSOC/17/J4.java
import java.util.*;
import java.io.*;
public class J4 {
	static class Comp implements Comparator<String> {
		public int compare(String a, String b) {
			if (map.get(a) > map.get(b))
				return -1;
			return 1;
		}
	}
	static HashMap<String, Integer> map;
	static List<String> key;
	public static void main (String[] args) throws IOException {
		map = new HashMap<String, Integer>();
		key = new ArrayList<String>();
		int N, C, M, num, total = 0;
		String arr[], str;
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		if (N < C) {
			System.out.println("Ace is dunzos.");
			return;
		}

		arr = br.readLine().toLowerCase().split(" ");
		num = Integer.parseInt(arr[arr.length - 1]);
		arr[arr.length - 1] = "";
		str = String.join("", arr);
		map.put(str, num);
		key.add(str);

		for (int i = 1; i < N; i++) {
			arr = br.readLine().toLowerCase().split(" ");
			num = Integer.parseInt(arr[arr.length - 1]);
			arr[arr.length - 1] = "";
			str = String.join("", arr);
			map.put(str, num);
			int loc = -Collections.binarySearch(key, str, Collections.reverseOrder()) - 1;
			key.add(loc, str);
		}
		num = 0;
		for (int i = 0; i < M; i++)	{
			str = br.readLine().toLowerCase();
			if (!map.containsKey(str)) {
				System.out.println("Ace is dunzos.");
				return;
			}
			num++;
			total += map.get(str);
			key.remove(str);
		}
		Collections.sort(key, new Comp());
		if (num < C || map.get(key.get(0)) > total/num) {
			for (int i = 0; (map.get(key.get(0)) >= total * 1.0 / num || i < C - num); i++) {
				total += map.get(key.get(0));
				num++;
				key.remove(0);
			}
		}
		if (total * 1.0 /num < 50)
			System.out.println("Ace is dunzos.");
		else
			System.out.format("%.2f\n", total * 1.0 /num);
	}
}

File: DMOJ/GFSSOC/17/J4.cpp
#include <stdio.h>
#include <utility>
#include <string.h>
#include <algorithm>
#include <unordered_map>

std::unordered_map<char*, int> map;
char grades[1000001][101];
char temp[101];
const char delim = ' ';

bool comp(char* a, char* b) {
	return map[a] > map[b];
}

int bin(int r, char* key) {
	int l = 0;
	while (l <= r) {
		int m = l + (r - l) / 2;
		if (strcmp(grades[m], key) < 0)
			r = m - 1;
		else if (strcmp(grades[m], key) > 0)
			l = m + 1;
		else if (strcmp(grades[m], key) == 0)
			return m;
	}
	return l;
}

int main() {
	int N, C, M, average = 0, last, num = 0;
	scanf("%d%d%d%100s%d", &N, &C, &M, temp, &last);
	
	memcpy(grades[0], temp, sizeof(temp) / sizeof(temp[0]));
	map[temp] = last;

	if (N < C) {
		printf("Ace is dunzos.\n");
		return 0;
	}

	for (int i = 1; i < N; i++) {
		scanf("%100s%d", temp, &last);
		int loc = bin(i, temp);
		memmove(grades[loc] + sizeof(grades[loc]), grades[loc], i - loc);
		memcpy(grades[loc], temp, sizeof(temp) / sizeof(temp[0]));
		map[temp] = last;
	}

	for (int i = 0; i < M; i++) {
		scanf("%100s", temp);
		if (!std::binary_search(grades[0], grades[0] + N, temp)) {
			printf("Ace is dunzos.\n");
			return 0;
		}
		average += map[temp];
		map[temp] = 0;
		num ++;
	}
	
	std::sort(grades[0], grades[0] + sizeof(grades) / sizeof(grades[0]), comp);

	if (M < C || map[grades[N]] > average * 1.0 / num) {
		for (int i = N - 1; i >= 0 || map[grades[N]] > average * 1.0 / num; i--) {
			average += map[grades[i]];
			num++;
		}
	}

	if (average * 1.0 / num >= 50)
		printf("%.2f\n", average * 1.0 / C);
	else
		printf("Ace is dunzos.\n");
}
